### PR TITLE
Option to inject kafka metadata

### PR DESCRIPTION
**Pull request by tabmatfournier**
- adds KafkaMetadataTransform to inject metadata to Struct/Map records 
- adds option to inject Kafka metadata to both DMS and Debezium SMTs w/o an additional record copy (using an internally nested KafkaMetadataAppender) 
- can be configured to inject an (optional) static field such as the (external) kafka cluster address/name

### CODE REVIEW

File: kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/DebeziumTransform.java [hunk 1]
```diff
@@ -110,6 +133,12 @@ private R applyWithSchema(R record) {
     }
     newValue.put(CdcConstants.COL_CDC, cdcMetadata);
 
+    if (kafkaAppender != null) {
+      if (record instanceof SinkRecord) {
```
Comment (tabmatfournier) [inline kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/DebeziumTransform.java:137]: Don't love these casts but type `R` is erased at runtime and it's hard to get even via reflection. 

The better solution would be to separate this into concrete Source/SinkRecord classes but that would be a breaking change.

On a `Source` record there really is no point to injecting metadata.

File: kafka-connect-transforms/src/test/java/io/tabular/iceberg/connect/transforms/util/KafkaMetadataAppenderTest.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.kafka.common.record.TimestampType;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class KafkaMetadataAppenderTest {
```
Comment (tabmatfournier) [inline kafka-connect-transforms/src/test/java/io/tabular/iceberg/connect/transforms/util/KafkaMetadataAppenderTest.java:34]: reworked this last night, I have a few more tests coming for the flattening functionality that I added.

File: kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/DmsTransform.java [hunk 3]
```diff
@@ -82,6 +110,8 @@ private R applySchemaless(R record) {
     Map<String, Object> newValue = Maps.newHashMap((Map<String, Object>) dataObj);
     newValue.put(CdcConstants.COL_CDC, cdcMetadata);
 
+    this.kafkaAppender.addToMap(record, newValue);
```
[the remark this answers is not in the document]
Comment (tabmatfournier) [reply kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/DmsTransform.java:113]: I can pull them all out.  Makes it easier since they capture ConnectRecord rather than SinkClass

File: kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/util/KafkaMetadataAppender.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+        @Override
+        public Struct addToStruct(R record, Struct struct) {
+          Struct nested = new Struct(nestedSchema);
+          nested.put(topicFieldName, record.topic());
+          nested.put(partitionFieldName, record.kafkaPartition());
+          if (record instanceof SinkRecord) {
+            SinkRecord sinkRecord = (SinkRecord) record;
+            nested.put(offsetFieldName, sinkRecord.kafkaOffset());
+          }
```
[the remark this answers is not in the document]
Comment (tabmatfournier) [reply kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/util/KafkaMetadataAppender.java:120]: :+1: Will remove from the Cdc/DMS transforms and make it just use `SinkRecord` as a standalone SMT

File: kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/util/KafkaMetadataAppender.java [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+              ConfigDef.Type.BOOLEAN,
+              true,
+              ConfigDef.Importance.LOW,
+              "Include appending of Kafka metadata to SinkRecord")
+          .define(
+              KafkaMetadataAppender.KEY_METADATA_FIELD_NAME,
+              ConfigDef.Type.STRING,
+              KafkaMetadataAppender.DEFAULT_METADATA_FIELD_NAME,
+              ConfigDef.Importance.LOW,
+              "field to append Kafka metadata under")
```
[the remark this answers is not in the document]
Comment (tabmatfournier) [reply kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/util/KafkaMetadataAppender.java:54]: Not sure what the question is?

[the remark this answers is not in the document]
Comment (tabmatfournier) [reply kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/util/KafkaMetadataAppender.java:54]: Oh.  I see.  That would break things. 

It's theoretically possible to make that happen --if the existing field was a struct but there is no guarantee.

At some point users have to be responsible for their own actions.

File: kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/KafkaMetadataTransform.java [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+
+public class KafkaMetadataTransform implements Transformation<SinkRecord> {
+
+  private interface RecordAppender {
+
+    SchemaBuilder addToSchema(SchemaBuilder builder);
+
+    Struct addToStruct(SinkRecord record, Struct struct);
+
+    Map<String, Object> addToMap(SinkRecord record, Map<String, Object> map);
```
Comment (fqtab) [inline kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/KafkaMetadataTransform.java:44]:
```suggestion
    void addToSchema(SchemaBuilder builder);

    void addToStruct(SinkRecord record, Struct struct);

    void addToMap(SinkRecord record, Map<String, Object> map);
```

nit: I would just make them all return void since these are all mutating/side-effecting methods. You only use the return value for the first one anyway.

File: kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/util/KafkaMetadataAppender.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (fqtab) [reply kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/util/KafkaMetadataAppender.java]: If it breaks things today, I'm good with it 👍 

I don't think it would hurt to support such behaviour in the future, but I'm content to wait for the feature request before doing this.

File: kafka-connect-transforms/README.md [hunk 7]
```diff
@@ -24,6 +24,26 @@ The `CopyValue` SMT copies a value from one field to a new field.
 "transforms.copyId.target.field": "id_copy",
 ```
 
+# KafkaMetadataTransform
+_(Experimental)_
+
+The `KafkaMetadata` injects `topic`, `partition`, `offset`, `record timestamp`.
```
Comment (fqtab) [inline kafka-connect-transforms/README.md:30]:
```suggestion
The `KafkaMetadata` injects `topic`, `partition`, `offset`, `record_timestamp`.
```

Although I'm still pushing for just `timestamp`.

**Review by fqtab — COMMENTED** [review]
Quite close, think we can wrap this up today

File: kafka-connect-transforms/README.md [hunk 8]
```diff
@@ -47,77 +67,3 @@ It will promote the `before` or `after` element fields to top level and add the
...
-SinkRecord.schema: 
-  "key": (Optional) Int32,
-  "array": (Optional) Array<String>,
-  "nested_object": (Optional) Map<string, String>
-  
-SinkRecord.value (Struct):
- "key" 1, 
- "array" ["1", "two", "3"] 
- "nested_object" Map ("some_key" : "["one", "two"]") 
-```
```
[the remark this answers is not in the document]
Comment (tabmatfournier) [reply kafka-connect-transforms/README.md:123]: 🤦  bad rebase

File: kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/KafkaMetadataTransform.java [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+
+    if (nestedMetadata) {
+      externalFieldAppender =
+          getExternalFieldAppender(config.getString(EXTERNAL_KAFKA_METADATA), name -> name);
+
+      SchemaBuilder nestedSchemaBuilder = SchemaBuilder.struct();
+      nestedSchemaBuilder
+          .field(TOPIC, Schema.STRING_SCHEMA)
+          .field(PARTITION, Schema.INT32_SCHEMA)
+          .field(OFFSET, Schema.OPTIONAL_INT64_SCHEMA)
```
[the remark this answers is not in the document]
Comment (tabmatfournier) [reply kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/KafkaMetadataTransform.java:155]: Right.  Didn't catch that after making SinkRecord concrete.

File: kafka-connect-transforms/README.md (hunk 7 above)
Comment (tabmatfournier) [reply kafka-connect-transforms/README.md:30]: Updated the code but forgot this esction of the docs, will fix.

File: kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/KafkaMetadataTransform.java [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+        return map;
+      }
+    };
+  }
+
+  private static final String TOPIC = "topic";
+  private static final String PARTITION = "partition";
+
+  private static final String OFFSET = "offset";
+
```
[the remark this answers is not in the document]
Comment (tabmatfournier) [reply kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/KafkaMetadataTransform.java:101]: Huh.  Would have expected spotless to have applied this

**Comment by tabmatfournier** [comment]
> Quite close, think we can wrap this up today. The only major comment is I don't think offset field should be optional.

Addressed all of this.

File: kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/KafkaMetadataTransform.java [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+      externalFieldAppender =
+          getExternalFieldAppender(config.getString(EXTERNAL_KAFKA_METADATA), namer);
+      return new RecordAppender() {
+        @Override
+        public void addToSchema(SchemaBuilder builder) {
+          builder
+              .field(topicFieldName, Schema.STRING_SCHEMA)
+              .field(partitionFieldName, Schema.INT32_SCHEMA)
+              .field(offsetFieldName, Schema.OPTIONAL_INT64_SCHEMA)
```
Comment (fqtab) [inline kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/KafkaMetadataTransform.java:198]:
```suggestion
              .field(offsetFieldName, Schema.INT64_SCHEMA)
```

this is still optional

File: kafka-connect-transforms/README.md [hunk 12]
```diff
@@ -120,4 +141,4 @@ SinkRecord.value (Struct):
  "key" 1, 
  "array" ["1", "two", "3"] 
  "nested_object" Map ("some_key" : "["one", "two"]") 
-```
+```
```
Comment (fqtab) [inline kafka-connect-transforms/README.md:144]: let's avoid unnecessary whitespace changes

File: kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/KafkaMetadataTransform.java [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+    SchemaBuilder builder = SchemaUtil.copySchemaBasics(schema, SchemaBuilder.struct());
+    for (Field field : schema.fields()) {
+      builder.field(field.name(), field.schema());
+    }
+    recordAppender.addToSchema(builder);
+    return builder.build();
+  }
+
+  private SinkRecord applySchemaless(SinkRecord record) {
+    Map<String, Object> value = Requirements.requireMap(record.value(), "KafkaMetadata transform");
```
Comment (fqtab) [inline kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/KafkaMetadataTransform.java:268]:
```suggestion
    Map<String, Object> value = Requirements.requireMap(record.value(), "KafkaMetadataTransform");
```

**Review by fqtab — APPROVED** [review]
